### ISSUE TITLE
Support redirect for twisted.web.client.Agent

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5157]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#5157|
|Type|enhancement|
|Created|2011-06-25 10:31:34Z|
|Branch|https://github.com/twisted/twisted/tree/agent-redirect-5157|

We need to support redirects, probably through a new RedirectAgent class.

<details><summary>Searchable metadata</summary>

```
trac-id__5157 5157
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_agent_redirect_5157 branches/agent-redirect-5157
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__web web
keywords__httpclient httpclient
time__1308997894000000 1308997894000000
changetime__1313827565000000 1313827565000000
version__None None
owner__therve therve
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [32178]) Branching to 'agent-redirect-5157'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

I've made a first implementation of RedirectAgent in the branch, but I think I need some feedback.

The implementation is fairly strict (treating 301 and 302 as 307). I'm returning a fair number of ResponseFailed, which I modified to include the response in case the redirect responsibility is given to the caller. Other than that I'm handling redirect limit rather easily.

I've also been wondering about keeping track of 301 to not query it the second time the same agent gets a request againts a permanently redirected URI.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

How does the branch handle redirect responses on requests that include bodies?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

For now, the branch ignores the body on the subsequent request after redirect (ie, pass the body the first time, and pass None afterwards). I failed to find something in the RFC about it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

Actually, I implemented that:

```
The action required MAY be carried out by the user agent without interaction with the user if and  only if the method used in the second request is GET or HEAD.
```

IE, the redirected request will only ever be HEAD or GET, so won't contain a body.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

  1. `twisted.web.client.RedirectLimitReached` looks like it means the same thing as `twisted.web.error.InfiniteRedirection`.  `InfiniteRedirection` looks like it offers some nice attributes, too.  Can we avoid introducing the latter and just re-use the former?
  1. Likewise for `client.NoAutomaticRedirect` / `error.PageRedirect`.
  1. The other new exception type, `RedirectWithNoLocation`, should probably be defined in `twisted.web.error` as well.
  1. `twisted.web.http` defines symbolic names for 301, 302, 303, 307 - use those instead of the literal integers in `_handleResponse`.

I didn't get a chance to finish this, I'll try to look again soon if no one else beats me to it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

Okay, I suppose I did mostly finish.  The only other points are that the news fragment is missing and that the http client howto should be updated with an example for `RedirectAgent`. :)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

Almost there, just need to update the documentation.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve removed owner|
|-|-|

Everything handled, I think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @therve|
|-|-|

Looks good to me.  The tests could perhaps use the response code mnemonics instead of the literal codes (we have them all defined in `twisted.web.http` I believe).  However, it's low level code for the handling of particular response codes, so having the literal values doesn't bother me too much.  Up to you.

Buildbot [likes it](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/agent-redirect-5157).  Please merge when you're happy with it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

I wondered about the tests. But for example we're using 200 almost everywhere, so I don't think using the literal values is that bad. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set status to `closed`|
|-|-|

(In [32466]) Merge agent-redirect-5157

Author: therve
Reviewer: exarkun
Fixes: [#5157](https://github.com/twisted/twisted/issues/5157)

Add a new twisted.web.client.RedirectAgent for redirect support the HTTP 1.1
client stack. The branch also makes the response available in ResponseFailed
errors.
